### PR TITLE
Core data - implemented core data to the project

### PR DESCRIPTION
**Pull request by dnikolic98**

### CODE REVIEW

File: weatherApp/App/Data/Core Data/CoreDataStack.swift [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+//  Created by Dario Nikolic on 19/08/2020.
+//  Copyright © 2020 Dario Nikolic. All rights reserved.
+//
+
+import CoreData
+
+class CoreDataStack {
+    
+    public typealias CoreDataManagerCompletion = () -> ()
+    public static let weatherApp = "weatherApp"
```
Comment (labasbozidar) [inline weatherApp/App/Data/Core Data/CoreDataStack.swift]: If you want to make this class "generic" and reusable outside of this app you need to remove weatherApp static property.

File: weatherApp/App/Data/Core Data/CoreDataStack.swift [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }()
+
+    private lazy var persistentStoreCoordinator: NSPersistentStoreCoordinator = {
+        return NSPersistentStoreCoordinator(managedObjectModel: self.managedObjectModel)
+    }()
+    
+    // MARK: - Public API
+    
+    public func saveChanges() {
+        mainManagedObjectContext.performAndWait {
```
Comment (labasbozidar) [inline weatherApp/App/Data/Core Data/CoreDataStack.swift]: Can you try to create one function for this performAndWait with context param? You could reuse this func two times.

File: weatherApp/App/Data/WeatherRepository.swift [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+//  Copyright © 2020 Dario Nikolic. All rights reserved.
+//
+
+import CoreData
+import Reachability
+
+class WeatherRepository {
+    
+    let weatherService: WeatherServiceProtocol
+    let coreDataService: CoreDataService
```
Comment (labasbozidar) [inline weatherApp/App/Data/WeatherRepository.swift]: You can create a protocol for CoreDataService also

File: weatherApp/App/Data/WeatherRepository.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    
+    
+    func fetchSeveralCurrentWeather(id: [Int], completion: @escaping (([CurrentWeatherCD]?) -> Void)) {
+        switch reachability.connection {
+        case .unavailable:
+            let currentWeatherListCD = coreDataService.fetchCurrentWeather()
+            completion(currentWeatherListCD)
+        default:
+            self.weatherService.fetchSeveralCurrentWeather(id: id) { (currentWeatherList) in
```
Comment (labasbozidar) [inline weatherApp/App/Data/WeatherRepository.swift]: You don't need to use self here.

File: weatherApp/App/Data/WeatherRepository.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                guard let currentWeatherList = currentWeatherList else {
+                    completion(nil)
+                    return
+                }
+                let _ = currentWeatherList.map { self.coreDataService.createCurrentWeatherFrom(currentWeather: $0) }
+                self.coreDataService.saveChanges()
```
Comment (labasbozidar) [inline weatherApp/App/Data/WeatherRepository.swift:41]: This is async block. You should use weak self.

File: weatherApp/App/Data/WeatherRepository.swift [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            let currentWeatherListCD = coreDataService.fetchForecastWeather(coord: coord)
+            completion(currentWeatherListCD)
+        default:
+            self.weatherService.fetchForcastWeather(coord: coord) { (forecastedWeather) in
+                guard let forecastedWeather = forecastedWeather else {
+                    completion(nil)
+                    return
+                }
+                
+                let _ = self.coreDataService.createForecastedWeatherFrom(forecastedWeather: forecastedWeather)
```
Comment (labasbozidar) [inline weatherApp/App/Data/WeatherRepository.swift:64]: Weak self

File: weatherApp/App/Data/WeatherRepository.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                    let currentWeatherListCD = self.coreDataService.fetchForecastWeather(coord: coord)
+                    completion(currentWeatherListCD)
+                }
+            }
+        }
+    }
+    
+    private func startReachability() {
+        do {
+            try reachability.startNotifier()
```
Comment (labasbozidar) [inline weatherApp/App/Data/WeatherRepository.swift:75]: Are you sure that you need this?

**Review by labasbozidar — CHANGES_REQUESTED** [review]

File: weatherApp/App/Data/WeatherRepository.swift (hunk 7 above)
Comment (dnikolic98) [reply weatherApp/App/Data/WeatherRepository.swift:75]: Its needed for making Reachability work properly.
Im planning to remove Reachability entirely soon and replace it with completion handling from NetworkService response.

File: weatherApp/App/Data/Core Data/CoreDataService.swift [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    init(coreDataStack: CoreDataStackProtocol) {
+        self.coreDataStack = coreDataStack
+        self.mainContext = coreDataStack.mainManagedObjectContext
+        self.privateContext = coreDataStack.privateChildManagedObjectContext()
+    }
+    
+    //MARK: - Fetches
+    
+    func fetchCurrentWeather() -> [CurrentWeatherCD]? {
+
```
Comment (istaudacher) [inline weatherApp/App/Data/Core Data/CoreDataService.swift]: remove newline

File: weatherApp/App/Data/Core Data/CoreDataServiceProtocol.swift [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+//
+//  Created by Dario Nikolic on 25/08/2020.
+//  Copyright © 2020 Dario Nikolic. All rights reserved.
+//
+
+protocol CoreDataServiceProtocol {
+    
+    //MARK: - Fetches
+    
+    func fetchCurrentWeather() -> [CurrentWeatherCD]?
```
Comment (istaudacher) [inline weatherApp/App/Data/Core Data/CoreDataServiceProtocol.swift]: A good API returns an empty array, not an optional when the return type is an array.

File: weatherApp/App/Data/WeatherRepository.swift [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                    let self = self,
+                    let currentWeatherList = currentWeatherList
+                else {
+                    completion(nil)
+                    return
+                }
+                let _ = currentWeatherList.map { self.coreDataService.createCurrentWeatherFrom(currentWeather: $0) }
+                self.coreDataService.saveChanges()
+                
+                DispatchQueue.main.async {
```
Comment (istaudacher) [inline weatherApp/App/Data/WeatherRepository.swift]: Do not use `DispatchQueue.main.async` here. What if I want to call this repository from some service? I would block the main thread. The completion block should have that. Please change this in all places.

File: weatherApp/Core/Data/CoreDataStack.swift [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    
+    private func setupCoreDataStack() {
+        guard let persistentStoreCoordinator = mainManagedObjectContext.persistentStoreCoordinator else {
+            fatalError("Unable to Set Up Core Data Stack")
+        }
+        
+        DispatchQueue.global().async {
+            self.addPersistentStore(to: persistentStoreCoordinator)
+            
+            DispatchQueue.main.async { self.completion() }
```
Comment (istaudacher) [inline weatherApp/Core/Data/CoreDataStack.swift]: As above

File: weatherApp/Features/WeatherDetails/Data/CoreData Models/DailyTemperatureCD+CoreDataClass.swift [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+//
+//  Created by Dario Nikolic on 19/08/2020.
+//
+//
+
+import Foundation
+import CoreData
+
+@objc(DailyTemperatureCD)
+public class DailyTemperatureCD: NSManagedObject {
```
Comment (istaudacher) [inline weatherApp/Features/WeatherDetails/Data/CoreData Models/DailyTemperatureCD+CoreDataClass.swift]: Please do not use abbreviations as CD, use whole words (I presume CoreData?). I know that renaming CoreData entities can be painful ( :) ) but please try to do that.

File: weatherApp/Features/WeatherDetails/UI/DetailWeatherViewController.swift [hunk 13]
```diff
@@ -44,6 +44,7 @@ class DetailWeatherViewController: UIViewController {
     
     override func viewWillLayoutSubviews() {
         super.viewWillLayoutSubviews()
+print(FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask))
```
Comment (istaudacher) [inline weatherApp/Features/WeatherDetails/UI/DetailWeatherViewController.swift]: Move this to app delegate.

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: weatherApp/Features/WeatherDetails/Data/CoreData Models/ForecastedWeatherCoreData+CoreDataClass.swift [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+import Foundation
+import CoreData
+
+@objc(ForecastedWeatherCoreData)
+public class ForecastedWeatherCoreData: NSManagedObject {
+    
+    class func firstOrCreate(withLongitude lon: Double, withLatitude lat: Double, context: NSManagedObjectContext) -> ForecastedWeatherCoreData? {
+        let epsilon = 0.00001
+        let format = "longitude > %f AND longitude < %f AND latitude > %f AND latitude < %f"
+        let predicate = NSPredicate(format: format, lon-epsilon, lon+epsilon, lat-epsilon, lat+epsilon)
```
Comment (istaudacher) [inline weatherApp/Features/WeatherDetails/Data/CoreData Models/ForecastedWeatherCoreData+CoreDataClass.swift]: `lon - epsilon` `lon + epsilon` ...

File: weatherApp/Features/WeatherList/UI/HomeViewController.swift [hunk 15]
```diff
@@ -58,7 +58,7 @@ class HomeViewController: UIViewController {
     
     @objc private func bindViewModel() {
         currentWeatherListPresenter.fetchCurrentWeatherList() { (currentWeatherList) in
-            guard let _ = currentWeatherList else {
+            guard currentWeatherList.count != 0 else {
```
Comment (istaudacher) [inline weatherApp/Features/WeatherList/UI/HomeViewController.swift]: `guard !currentWeatherList.isEmpty else { return }`

**Review by istaudacher — CHANGES_REQUESTED** [review]

**Review by istaudacher — APPROVED** [review]

**Review by labasbozidar — APPROVED** [review]

**Pull request merged**